File: ccanvas.c
#include <stdint.h>
#include <stddef.h>
#include <stdlib.h>

#define PIXEL_SIZE 4  // RGBA

static uint8_t* framebuffer = NULL;
static size_t width = 0;
static size_t height = 0;
static size_t buffer_size = 0;

// Set resolution and allocate framebuffer
void set_resolution(size_t w, size_t h) {
    width = w;
    height = h;
    buffer_size = width * height * PIXEL_SIZE;

    if (framebuffer) {
        free(framebuffer);
    }

    framebuffer = (uint8_t*) malloc(buffer_size);
}

// Fill buffer with a gradient
void fill_gradient() {
    if (!framebuffer) return;

    for (size_t y = 0; y < height; y++) {
        for (size_t x = 0; x < width; x++) {
            size_t index = (y * width + x) * PIXEL_SIZE;
            framebuffer[index + 0] = x % 256;    // R
            framebuffer[index + 1] = y % 256;    // G
            framebuffer[index + 2] = 128;       // B
            framebuffer[index + 3] = 255;       // A
        }
    }
}

// Get pointer to framebuffer
uint8_t* get_framebuffer_ptr() {
    return framebuffer;
}

// Get size of framebuffer
size_t get_framebuffer_size() {
    return buffer_size;
}

File: ccanvas.js
const ccanvasTemplate = `
<canvas width="320" height="240"></canvas>
`;


class CCanvas extends HTMLElement {

    static observedAttributes=['width', 'height', 'background-color'];
    #canvas;
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
        this.shadowRoot.innerHTML = ccanvasTemplate;
        this.#canvas = this.shadowRoot.querySelector('canvas');

        this.#init();
    }

    async #init() {
        const response = await fetch('ccanvas.wasm');
        const bytes = await response.arrayBuffer();
        const { instance } = await WebAssembly.instantiate(bytes, {});
      
        const { memory, fill_gradient, get_framebuffer_ptr, get_framebuffer_size, set_resolution} = instance.exports;
      

        const w = 320;
        const h = 240;

        this.width = 320;
        this.height = h;
        // Call C function to fill the buffer
        set_resolution(w, h);
        fill_gradient();
      
        const ptr = get_framebuffer_ptr();
        const size = get_framebuffer_size();
        const buffer = new Uint8ClampedArray(memory.buffer, ptr, size);
      
        const ctx = this.#canvas.getContext('2d');
        const imageData = new ImageData(buffer, this.width, this.height);
        ctx.putImageData(imageData, 0, 0);
    }

    connectedCallback() {
    }

    attributeChangedCallback(name, oldValue, newValue) {

        switch (name) {
            case 'width':
                this.width = newValue;
                break;
            case 'height':
                this.height = newValue;
            break;
            case 'background-color':
                this.#canvas.style.backgroundColor = newValue;
                break;
            default:
                break;
        }
    }

    get width() {
        return this.#canvas.width;
    }
    set width(val) {
        this.#canvas.width = val;
    }
    get height() {
        return this.#canvas.height;
    }
    set height(val) {
        this.#canvas.height = val;
    }
}

customElements.define('c-canvas', CCanvas);